analysis/estimate/scripts/prepare-expression-data.R

```diff
@@ -0,0 +1,206 @@
+id_map_df <- readr::read_tsv(opts$ensembl_symbol_map_file, show_col_types = FALSE)
+
+
+
+# Prepare TPM counts data frame ----------------
+
+t2g_table <- readr::read_tsv(
+  opts$t2g_file,
+  col_names = c("transcript_id", "gene_id"),
+  show_col_types = FALSE
+)
+
+# Find all quant.sf files 
+quant_files <- list.files(
+  path = opts$scpca_dir,
+  pattern = "quant.sf",
+  recursive = TRUE
+)
+stopifnot("Could not find any quant.sf files for the specified project." = length(quant_files) > 0)
+
+# Calculate TPM for each sample 
+sample_ids <- stringr::str_split_i(quant_files, pattern = "/", i = 1)
+quant_paths <- setNames(file.path(opts$scpca_dir, quant_files), sample_ids)
+
+tpm_df <- tximport::tximport(
+  quant_paths,
+  type = "salmon",
+  tx2gene = t2g_table
+) |>
+  purrr::pluck("abundance") |>
+  as.data.frame() |>
+  tibble::rownames_to_column("ensembl_id") |>
+  convert_ids(id_map_df)
+
+tpm_samples <- colnames(tpm_df[,-1]) # used to prepare bulk counts
+
+# Prepare bulk counts data frame ----------------
+
+bulk_file <- file.path(
+  opts$scpca_dir, 
+  glue::glue("{opts$project_id}_bulk_quant.tsv")
+)
+raw_bulk_df <- readr::read_tsv(bulk_file, show_col_types = FALSE) 
+
+# read library metadata to swap library to sample ids
+sample_library_df <- readr::read_tsv(opts$library_metadata_file, show_col_types = FALSE) |>
+  dplyr::filter(
+    scpca_project_id == opts$project_id,
+    seq_unit == "bulk"
+  ) |> 
+  dplyr::select(scpca_sample_id, scpca_library_id)
+
+bulk_df <- raw_bulk_df |>
+  dplyr::rename(ensembl_id = gene_id) |>
+  # replace library with sample ids
+  tidyr::pivot_longer(contains("SCPCL")) |>
+  dplyr::left_join(sample_library_df, by = c("name" = "scpca_library_id")) |>
+  dplyr::select(-name) |>
+  # subset to only bulk samples in the TPM df, since bulk raw counts may have more columns than we want
+  dplyr::filter(scpca_sample_id %in% tpm_samples) |>
+  tidyr::pivot_wider(
+    names_from = scpca_sample_id, 
+    values_from = value
+  ) |>
+  convert_ids(id_map_df)
+
+# Prepare pseudobulk counts data frame ----------------
+
+# read all sces
+rds_files <- list.files(
+  path = opts$scpca_dir,
+  pattern = "_processed\\.rds$",
+  recursive = TRUE
+)
+stopifnot(
+  "Could not find any _processed.rds files in the provided `input_dir`." = length(rds_files) > 0
+)
+sample_ids <- stringr::str_split_i(rds_files, pattern = "/", i = 1)
+rds_files <- file.path(opts$scpca_dir, rds_files) # update to contain full path
+names(rds_files) <- sample_ids
+sce_list <- purrr::map(rds_files, readr::read_rds)
+
+# extract and sum the raw counts
+pseudo_df <- sce_list |>
+  purrr::map(counts) |>
+  purrr::map(rowSums) |>
+  do.call(cbind, args = _ ) |>
+  as.data.frame() |>
+  tibble::rownames_to_column("ensembl_id") |>
+  convert_ids(id_map_df)
+
+
+
+# Export RDS -----------------
+
+expr_list <- list(
+  "bulk_counts" = bulk_df, 
+  "bulk_tpm" = tpm_df, 
+  "pseudobulk" = pseudo_df
+)
+# check that all data frames are compatible:
```

Comment:
Is this comment meant to be here?

Reply:
It is, to preface the next lines which check that they have the same genes (at least the two bulks) and all have the same samples, but the spacing is unclear I think so I'll tweak that!